Source/Particles/PhysicalParticleContainer.cpp

```diff
@@ -3122,107 +3129,163 @@ PhysicalParticleContainer::ImplicitPushXP (WarpXParIter& pti,
-        amrex::ParticleReal Ezp = Ez_external_particle;
-        amrex::ParticleReal Bxp = Bx_external_particle;
-        amrex::ParticleReal Byp = By_external_particle;
-        amrex::ParticleReal Bzp = Bz_external_particle;
+        amrex::ParticleReal step_norm = 1._prt;
+        for (int iter=0; iter<max_iterations;) {
 
-        if(!t_do_not_gather){
-            // first gather E and B to the particle positions
-            doGatherShapeNImplicit(xp_n, yp_n, zp_n, xp, yp, zp, Exp, Eyp, Ezp, Bxp, Byp, Bzp,
-                                   ex_arr, ey_arr, ez_arr, bx_arr, by_arr, bz_arr,
-                                   ex_type, ey_type, ez_type, bx_type, by_type, bz_type,
-                                   dx_arr, xyzmin_arr, lo, n_rz_azimuthal_modes, nox,
-                                   depos_type );
-        }
+            dxp = 0.0;
+            dyp = 0.0;
+            dzp = 0.0;
+            UpdatePositionImplicit(dxp, dyp, dzp, ux_n[ip], uy_n[ip], uz_n[ip], ux[ip], uy[ip], uz[ip], 0.5_rt*dt);
+#if !defined(WARPX_DIM_1D_Z)
+            xp = xp_n + dxp;
+#endif
+#if defined(WARPX_DIM_3D) || defined(WARPX_DIM_RZ)
+            yp = yp_n + dyp;
+#endif
+            zp = zp_n + dzp;
+            setPosition(ip, xp, yp, zp);
 
-        // Externally applied E and B-field in Cartesian co-ordinates
-        [[maybe_unused]] const auto& getExternalEB_tmp = getExternalEB;
-        if constexpr (exteb_control == has_exteb) {
-            getExternalEB(ip, Exp, Eyp, Ezp, Bxp, Byp, Bzp);
-        }
+            PositionNorm( dxp, dyp, dzp, dxp_save, dyp_save, dzp_save,
+                          idxg2, idyg2, idzg2, step_norm, iter );
+            if( step_norm < particle_tolerance ) { break; }
+
+            amrex::ParticleReal Exp = Ex_external_particle;
+            amrex::ParticleReal Eyp = Ey_external_particle;
+            amrex::ParticleReal Ezp = Ez_external_particle;
+            amrex::ParticleReal Bxp = Bx_external_particle;
+            amrex::ParticleReal Byp = By_external_particle;
+            amrex::ParticleReal Bzp = Bz_external_particle;
+
+            if(!t_do_not_gather){
+                // first gather E and B to the particle positions
+                doGatherShapeNImplicit(xp_n, yp_n, zp_n, xp, yp, zp, Exp, Eyp, Ezp, Bxp, Byp, Bzp,
+                                       ex_arr, ey_arr, ez_arr, bx_arr, by_arr, bz_arr,
+                                       ex_type, ey_type, ez_type, bx_type, by_type, bz_type,
+                                       dx_arr, xyzmin_arr, lo, n_rz_azimuthal_modes, nox,
+                                       depos_type );
+            }
 
-        scaleFields(xp, yp, zp, Exp, Eyp, Ezp, Bxp, Byp, Bzp);
+            // Externally applied E and B-field in Cartesian co-ordinates
+            [[maybe_unused]] const auto& getExternalEB_tmp = getExternalEB;
+            if constexpr (exteb_control == has_exteb) {
+                getExternalEB(ip, Exp, Eyp, Ezp, Bxp, Byp, Bzp);
```

Comment:
Note to self: check that this uses the right time, to evaluate the external E and B fields.